server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java

```diff
@@ -380,30 +372,87 @@ private static long computeMatchingBytes(TransportNodesListShardStoreMetaData.St
         }
     }
 
+    private static long getRetainingSeqNoForNode(TransportNodesListShardStoreMetaData.StoreFilesMetaData primaryStore, DiscoveryNode node) {
+        final String retentionLeaseId = ReplicationTracker.getPeerRecoveryRetentionLeaseId(node.getId());
+        return primaryStore.peerRecoveryRetentionLeases().stream()
+            .filter(lease -> lease.id().equals(retentionLeaseId))
+            .mapToLong(RetentionLease::retainingSequenceNumber).findFirst().orElse(0L);
+    }
+
+    private static MatchingNode computeMatchingNode(
+        DiscoveryNode primaryNode, TransportNodesListShardStoreMetaData.StoreFilesMetaData primaryStore,
+        DiscoveryNode replicaNode, TransportNodesListShardStoreMetaData.StoreFilesMetaData replicaStore) {
+        if (replicaStore.isEmpty()) {
+            return new MatchingNode(0, 0, false); // store is corrupted
+        }
+        final long retainingSeqNoForPrimary = getRetainingSeqNoForNode(primaryStore, primaryNode);
+        final long retainingSeqNoForReplica = getRetainingSeqNoForNode(primaryStore, replicaNode);
+        final long matchingBytes = computeMatchingBytes(primaryStore, replicaStore);
+        final boolean isNoopRecovery = matchingBytes == Long.MAX_VALUE
```

Comment:
can you change `computeMatchingBytes` not to do the sync id check anymore, but do that one here instead? That means that `matchingBytes` will truly mean the number of exact matching bytes.